File: prototype/js/grammar-b.mjs
import { LIMITS } from "./constants.mjs";
import { decodeVarUint, encodeVarUint } from "./varint.mjs";

export const GRAMMAR_B_VERSION = 1;

export const PACKED_OPCODE = Object.freeze({
  END: 0,
  SKIP: 1,
  LITERAL: 2,
  REPEAT_TOKEN: 3,
  SET_GLYPH: 4,
  SET_FOREGROUND: 5,
  SET_BACKGROUND: 6,
  SET_COLOR_PAIR: 7
});

const OPCODE_NAMES = Object.freeze(
  Object.fromEntries(Object.entries(PACKED_OPCODE).map(([name, value]) => [value, name]))
);

function checkedGrid(columns, rows) {
  if (!Number.isInteger(columns) || !Number.isInteger(rows) || columns < 1 || rows < 1 ||
      columns > LIMITS.maxColumns || rows > LIMITS.maxRows) {
    throw new RangeError("Grid exceeds V64 proof profile");
  }
  const cellCount = columns * rows;
  if (!Number.isSafeInteger(cellCount) || cellCount > LIMITS.maxCells) {
    throw new RangeError("Cell count exceeds V64 proof profile");
  }
  return cellCount;
}

export function paletteIndexBits(paletteDepth) {
  if (!Number.isInteger(paletteDepth) || paletteDepth < 2 || paletteDepth > 256) {
    throw new RangeError("Palette depth must be an integer from 2 through 256");
  }
  return Math.ceil(Math.log2(paletteDepth));
}

function equalToken(a, aOffset, b, bOffset) {
  return a[aOffset] === b[bOffset] &&
    a[aOffset + 1] === b[bOffset + 1] &&
    a[aOffset + 2] === b[bOffset + 2];
}

function isVoidToken(cells, offset) {
  return cells[offset] === 0 && cells[offset + 1] === 0 && cells[offset + 2] === 0;
}

function tokenAt(cells, cell) {
  const offset = cell * 3;
  return [cells[offset], cells[offset + 1], cells[offset + 2]];
}

function validateToken(token, paletteDepth) {
  if (!Array.isArray(token) || token.length !== 3) throw new Error("Malformed cell token");
  const [glyph, foreground, background] = token;
  if (!Number.isInteger(glyph) || glyph < 0 || glyph >= 64) throw new RangeError("Glyph index exceeds canonical set");
  if (!Number.isInteger(foreground) || !Number.isInteger(background) ||
      foreground < 0 || background < 0 ||
      foreground >= paletteDepth || background >= paletteDepth) {
    throw new RangeError("Palette index exceeds declared depth");
  }
}

function changeKind(current, previous, cell) {
  const offset = cell * 3;
  const glyph = current[offset] !== previous[offset];
  const foreground = current[offset + 1] !== previous[offset + 1];
  const background = current[offset + 2] !== previous[offset + 2];
  if (glyph && !foreground && !background) return "SET_GLYPH";
  if (!glyph && foreground && !background) return "SET_FOREGROUND";
  if (!glyph && !foreground && background) return "SET_BACKGROUND";
  if (!glyph && foreground && background) return "SET_COLOR_PAIR";
  return glyph || foreground || background ? "LITERAL" : "SKIP";
}

function countEqualTokens(current, previous, start, cellCount, keyframe) {
  const offset = start * 3;
  let count = 1;
  while (start + count < cellCount &&
      equalToken(current, offset, current, (start + count) * 3) &&
      (keyframe || !equalToken(current, (start + count) * 3, previous, (start + count) * 3))) {
    count += 1;
  }
  return count;
}

export function buildCommandTrace(currentInput, previousInput, options) {
  const { columns, rows, paletteDepth, keyframe = !previousInput } = options;
  const cellCount = checkedGrid(columns, rows);
  const current = Uint8Array.from(currentInput);
  if (current.length !== cellCount * 3) throw new RangeError("Current cell-state length mismatch");
  const previous = keyframe ? new Uint8Array(current.length) : Uint8Array.from(previousInput || []);
  if (!keyframe && previous.length !== current.length) {
    throw new RangeError("Delta frame requires a matching previous state");
  }
  for (let cell = 0; cell < cellCount; cell += 1) validateToken(tokenAt(current, cell), paletteDepth);

  const commands = [];
  let cell = 0;
  while (cell < cellCount) {
    const offset = cell * 3;
    const unchanged = keyframe ? isVoidToken(current, offset) : equalToken(current, offset, previous, offset);
    if (unchanged) {
      const startCell = cell;
      cell += 1;
      while (cell < cellCount) {
        const at = cell * 3;
        const same = keyframe ? isVoidToken(current, at) : equalToken(current, at, previous, at);
        if (!same) break;
        cell += 1;
      }
      commands.push({ op: "SKIP", startCell, count: cell - startCell });
      continue;
    }

    const repeat = countEqualTokens(current, previous, cell, cellCount, keyframe);
    if (repeat >= 3) {
      commands.push({ op: "REPEAT_TOKEN", startCell: cell, count: repeat, token: tokenAt(current, cell) });
      cell += repeat;
      continue;
    }

    if (!keyframe) {
      const kind = changeKind(current, previous, cell);
      if (kind === "SET_GLYPH") {
        commands.push({ op: kind, startCell: cell, value: current[offset] });
        cell += 1;
        continue;
      }
      if (kind === "SET_FOREGROUND") {
        commands.push({ op: kind, startCell: cell, value: current[offset + 1] });
        cell += 1;
        continue;
      }
      if (kind === "SET_BACKGROUND") {
        commands.push({ op: kind, startCell: cell, value: current[offset + 2] });
        cell += 1;
        continue;
      }
      if (kind === "SET_COLOR_PAIR") {
        commands.push({
          op: kind,
          startCell: cell,
          foreground: current[offset + 1],
          background: current[offset + 2]
        });
        cell += 1;
        continue;
      }
    }

    const startCell = cell;
    const tokens = [tokenAt(current, cell)];
    cell += 1;
    while (cell < cellCount) {
      const at = cell * 3;
      const same = keyframe ? isVoidToken(current, at) : equalToken(current, at, previous, at);
      if (same || countEqualTokens(current, previous, cell, cellCount, keyframe) >= 3) break;
      if (!keyframe && changeKind(current, previous, cell) !== "LITERAL") break;
      tokens.push(tokenAt(current, cell));
      cell += 1;
    }
    commands.push({ op: "LITERAL", startCell, count: tokens.length, tokens });
  }
  commands.push({ op: "END", startCell: cell, count: 0 });
  return {
    grammar: "V64-GRAMMAR-B",
    version: GRAMMAR_B_VERSION,
    columns,
    rows,
    cellCount,
    paletteDepth,
    paletteBits: paletteIndexBits(paletteDepth),
    keyframe: Boolean(keyframe),
    commands
  };
}

function packFields(fields) {
  const totalBits = fields.reduce((sum, field) => sum + field.width, 0);
  const output = Buffer.alloc(Math.ceil(totalBits / 8));
  let bitOffset = 0;
  for (const { value, width } of fields) {
    const limit = 2 ** width;
    if (!Number.isInteger(value) || value < 0 || value >= limit) {
      throw new RangeError(`Value ${value} does not fit in ${width} bits`);
    }
    for (let bit = 0; bit < width; bit += 1) {
      if (value & (2 ** bit)) output[bitOffset >> 3] |= 1 << (bitOffset & 7);
      bitOffset += 1;
    }
  }
  return output;
}

function unpackFields(bytes, widths) {
  const totalBits = widths.reduce((sum, width) => sum + width, 0);
  const requiredBytes = Math.ceil(totalBits / 8);
  if (bytes.length !== requiredBytes) throw new Error("Packed field length mismatch");
  const values = [];
  let bitOffset = 0;
  for (const width of widths) {
    let value = 0;
    for (let bit = 0; bit < width; bit += 1) {
      if (bytes[bitOffset >> 3] & (1 << (bitOffset & 7))) value += 2 ** bit;
      bitOffset += 1;
    }
    values.push(value);
  }
  for (let bit = totalBits; bit < requiredBytes * 8; bit += 1) {
    if (bytes[bit >> 3] & (1 << (bit & 7))) throw new Error("Nonzero packed padding bits");
  }
  return values;
}

function packedTokens(tokens, paletteBits, paletteDepth) {
  const fields = [];
  for (const token of tokens) {
    validateToken(token, paletteDepth);
    fields.push(
      { value: token[0], width: 6 },
      { value: token[1], width: paletteBits },
      { value: token[2], width: paletteBits }
    );
  }
  return packFields(fields);
}

function pushVar(parts, value) {
  const encoded = encodeVarUint(value);
  parts.push(encoded);
  return encoded.length;
}

export function encodePackedCommands(trace) {
  if (trace?.grammar !== "V64-GRAMMAR-B" || trace.version !== GRAMMAR_B_VERSION) {
    throw new Error("Unsupported command trace");
  }
  const paletteBits = paletteIndexBits(trace.paletteDepth);
  const parts = [];
  for (const command of trace.commands) {
    const opcode = PACKED_OPCODE[command.op];
    if (opcode === undefined) throw new Error(`Unknown trace operation ${command.op}`);
    parts.push(Buffer.from([opcode]));
    if (command.op === "END") continue;
    if (command.op === "SKIP") {
      pushVar(parts, command.count);
    } else if (command.op === "LITERAL") {
      if (!command.count || command.tokens?.length !== command.count) throw new Error("Malformed literal trace");
      pushVar(parts, command.count);
      parts.push(packedTokens(command.tokens, paletteBits, trace.paletteDepth));
    } else if (command.op === "REPEAT_TOKEN") {
      if (!command.count) throw new Error("Malformed repeated-token trace");
      pushVar(parts, command.count);
      parts.push(packedTokens([command.token], paletteBits, trace.paletteDepth));
    } else if (command.op === "SET_GLYPH") {
      parts.push(packFields([{ value: command.value, width: 6 }]));
    } else if (command.op === "SET_FOREGROUND" || command.op === "SET_BACKGROUND") {
      if (command.value >= trace.paletteDepth) throw new RangeError("Palette index exceeds declared depth");
      parts.push(packFields([{ value: command.value, width: paletteBits }]));
    } else if (command.op === "SET_COLOR_PAIR") {
      if (command.foreground >= trace.paletteDepth || command.background >= trace.paletteDepth) {
        throw new RangeError("Palette index exceeds declared depth");
      }
      parts.push(packFields([
        { value: command.foreground, width: paletteBits },
        { value: command.background, width: paletteBits }
      ]));
    }
  }
  return Buffer.concat(parts);
}

export function parsePackedCommands(commandInput, options) {
  const commandBytes = Buffer.from(commandInput);
  const { columns, rows, paletteDepth, keyframe = false } = options;
  const cellCount = checkedGrid(columns, rows);
  const paletteBits = paletteIndexBits(paletteDepth);
  const tokenBits = 6 + 2 * paletteBits;
  const commands = [];
  const commandLimit = 2 * cellCount + 1;
  let offset = 0;
  let cursor = 0;
  let ended = false;

  const readVar = () => {
    const start = offset;
    const result = decodeVarUint(commandBytes, offset);
    offset = result.next;
    if (!result.value) throw new Error("Zero-progress command");
    return { value: result.value, bytes: offset - start };
  };
  const readPacked = (widths) => {
    const byteLength = Math.ceil(widths.reduce((sum, width) => sum + width, 0) / 8);
    if (offset + byteLength > commandBytes.length) throw new Error("Truncated packed command payload");
    const bytes = commandBytes.subarray(offset, offset + byteLength);
    offset += byteLength;
    return { values: unpackFields(bytes, widths), bytes: byteLength };
  };
  const push = (command) => {
    commands.push(command);
    if (commands.length > commandLimit) throw new Error("Frame command count exceeds bound");
  };

  while (offset < commandBytes.length) {
    const byteStart = offset;
    const opcode = commandBytes[offset++];
    const op = OPCODE_NAMES[opcode];
    if (!op) throw new Error(`Unknown mandatory packed opcode 0x${opcode.toString(16).padStart(2, "0")}`);
    if (op === "END") {
      push({
        op,
        startCell: cursor,
        count: 0,
        byteStart,
        byteLength: 1,
        opcodeBytes: 1,
        countBytes: 0,
        payloadBytes: 0
      });
      ended = true;
      break;
    }

    if (op === "SKIP") {
      const count = readVar();
      if (cursor + count.value > cellCount) throw new Error("Packed skip advances beyond grid");
      push({
        op,
        startCell: cursor,
        count: count.value,
        byteStart,
        byteLength: offset - byteStart,
        opcodeBytes: 1,
        countBytes: count.bytes,
        payloadBytes: 0
      });
      cursor += count.value;
      continue;
    }

    if (op === "LITERAL" || op === "REPEAT_TOKEN") {
      const count = readVar();
      if (cursor + count.value > cellCount) throw new Error("Packed token command advances beyond grid");
      const tokenCount = op === "LITERAL" ? count.value : 1;
      const widths = Array.from({ length: tokenCount }, () => [6, paletteBits, paletteBits]).flat();
      const packed = readPacked(widths);
      const tokens = [];
      for (let index = 0; index < packed.values.length; index += 3) {
        const token = packed.values.slice(index, index + 3);
        validateToken(token, paletteDepth);
        tokens.push(token);
      }
      push({
        op,
        startCell: cursor,
        count: count.value,
        ...(op === "LITERAL" ? { tokens } : { token: tokens[0] }),
        byteStart,
        byteLength: offset - byteStart,
        opcodeBytes: 1,
        countBytes: count.bytes,
        payloadBytes: Math.ceil(tokenCount * tokenBits / 8)
      });
      cursor += count.value;
      continue;
    }

    const widths = op === "SET_GLYPH" ? [6] :
      (op === "SET_COLOR_PAIR" ? [paletteBits, paletteBits] : [paletteBits]);
    const packed = readPacked(widths);
    if (cursor >= cellCount) throw new Error("Packed component command advances beyond grid");
    const values = packed.values;
    if (op !== "SET_GLYPH" && values.some((value) => value >= paletteDepth)) {
      throw new Error("Palette index exceeds declared depth");
    }
    push({
      op,
      startCell: cursor,
      count: 1,
      ...(op === "SET_COLOR_PAIR"
        ? { foreground: values[0], background: values[1] }
        : { value: values[0] }),
      byteStart,
      byteLength: offset - byteStart,
      opcodeBytes: 1,
      countBytes: 0,
      payloadBytes: packed.bytes
    });
    cursor += 1;
  }

  if (!ended) throw new Error("Packed command stream has no END");
  if (offset !== commandBytes.length) throw new Error("Trailing bytes after packed frame END");
  return {
    grammar: "V64-GRAMMAR-B",
    version: GRAMMAR_B_VERSION,
    columns,
    rows,
    cellCount,
    paletteDepth,
    paletteBits,
    keyframe: Boolean(keyframe),
    commands
  };
}

export function applyPackedCommands(commandBytes, priorInput, options) {
  const { columns, rows, paletteDepth, keyframe } = options;
  const cellCount = checkedGrid(columns, rows);
  if (!keyframe && (!priorInput || priorInput.length !== cellCount * 3)) {
    throw new Error("Delta frame has no valid prior state");
  }
  const state = keyframe ? new Uint8Array(cellCount * 3) : new Uint8Array(priorInput);
  const trace = parsePackedCommands(commandBytes, { columns, rows, paletteDepth, keyframe });
  for (const command of trace.commands) {
    if (command.op === "END" || command.op === "SKIP") continue;
    if (command.op === "LITERAL") {
      for (let index = 0; index < command.tokens.length; index += 1) {
        state.set(command.tokens[index], (command.startCell + index) * 3);
      }
    } else if (command.op === "REPEAT_TOKEN") {
      for (let index = 0; index < command.count; index += 1) {
        state.set(command.token, (command.startCell + index) * 3);
      }
    } else {
      const offset = command.startCell * 3;
      if (command.op === "SET_GLYPH") state[offset] = command.value;
      else if (command.op === "SET_FOREGROUND") state[offset + 1] = command.value;
      else if (command.op === "SET_BACKGROUND") state[offset + 2] = command.value;
      else if (command.op === "SET_COLOR_PAIR") {
        state[offset + 1] = command.foreground;
        state[offset + 2] = command.background;
      }
    }
  }
  return state;
}

export function measurePackedCommands(commandBytes, options) {
  const trace = parsePackedCommands(commandBytes, options);
  const result = {
    totalBytes: Buffer.byteLength(commandBytes),
    opcodeBytes: 0,
    countBytes: 0,
    payloadBytes: 0,
    opcodes: {},
    cells: {}
  };
  for (const command of trace.commands) {
    result.opcodeBytes += command.opcodeBytes;
    result.countBytes += command.countBytes;
    result.payloadBytes += command.payloadBytes;
    result.opcodes[command.op] = (result.opcodes[command.op] || 0) + 1;
    if (command.op !== "END") result.cells[command.op] = (result.cells[command.op] || 0) + command.count;
  }
  if (result.opcodeBytes + result.countBytes + result.payloadBytes !== result.totalBytes) {
    throw new Error("Packed command accounting mismatch");
  }
  return result;
}

File: crates/v64-core/src/grammar_b.rs
pub const END: u8 = 0;
pub const SKIP: u8 = 1;
pub const LITERAL: u8 = 2;
pub const REPEAT_TOKEN: u8 = 3;
pub const SET_GLYPH: u8 = 4;
pub const SET_FOREGROUND: u8 = 5;
pub const SET_BACKGROUND: u8 = 6;
pub const SET_COLOR_PAIR: u8 = 7;
pub const REPEAT_GLYPH: u8 = 8;
pub const REPEAT_FOREGROUND: u8 = 9;
pub const REPEAT_BACKGROUND: u8 = 10;
pub const REPEAT_COLOR_PAIR: u8 = 11;

/// Applies one direct Grammar B command stream transactionally.
///
/// The returned vector is the only committed state. The caller's prior state is
/// borrowed and therefore cannot be modified by a malformed partial stream.
pub fn apply_packed_commands(
    bytes: &[u8],
    prior: Option<&[u8]>,
    columns: usize,
    rows: usize,
    palette_depth: usize,
    keyframe: bool,
) -> Result<Vec<u8>, String> {
    let cells = columns
        .checked_mul(rows)
        .ok_or_else(|| "Cell count overflow".to_owned())?;
    if columns == 0
        || rows == 0
        || columns > usize::from(crate::MAX_COLUMNS)
        || rows > usize::from(crate::MAX_ROWS)
        || cells > crate::MAX_CELLS
    {
        return Err("Invalid or oversized Grammar B grid".to_owned());
    }
    let state_length = cells
        .checked_mul(3)
        .ok_or_else(|| "Cell-state length overflow".to_owned())?;
    let mut state = if keyframe {
        vec![0; state_length]
    } else {
        let prior = prior.ok_or_else(|| "Delta frame has no valid prior state".to_owned())?;
        if prior.len() != state_length {
            return Err("Delta frame has no valid prior state".to_owned());
        }
        prior.to_vec()
    };
    let palette_bits = palette_bits(palette_depth)?;
    let token_bits = 6 + 2 * palette_bits;
    let command_limit = cells
        .checked_mul(2)
        .and_then(|value| value.checked_add(1))
        .ok_or_else(|| "Frame command-count overflow".to_owned())?;
    let mut offset = 0usize;
    let mut cursor = 0usize;
    let mut command_count = 0usize;
    let mut ended = false;

    while offset < bytes.len() {
        command_count += 1;
        if command_count > command_limit {
            return Err("Frame command count exceeds bound".to_owned());
        }
        let opcode = bytes[offset];
        offset += 1;
        if opcode == END {
            ended = true;
            break;
        }
        if opcode == SKIP {
            let count = read_count(bytes, &mut offset)?;
            require_range(cursor, count, cells, "Packed skip advances beyond grid")?;
            cursor += count;
            continue;
        }
        if opcode == LITERAL {
            let count = read_count(bytes, &mut offset)?;
            require_range(
                cursor,
                count,
                cells,
                "Packed token command advances beyond grid",
            )?;
            let used_bits = count
                .checked_mul(token_bits)
                .ok_or_else(|| "Packed literal is too large".to_owned())?;
            let payload = read_payload(bytes, &mut offset, used_bits)?;
            let mut bit = 0usize;
            for index in 0..count {
                let glyph = read_bits(payload, bit, 6);
                bit += 6;
                let foreground = read_bits(payload, bit, palette_bits);
                bit += palette_bits;
                let background = read_bits(payload, bit, palette_bits);
                bit += palette_bits;
                write_token(
                    &mut state,
                    cursor + index,
                    glyph,
                    foreground,
                    background,
                    palette_depth,
                )?;
            }
            require_zero_padding(payload, used_bits)?;
            cursor += count;
            continue;
        }
        if opcode == REPEAT_TOKEN {
            let count = read_count(bytes, &mut offset)?;
            require_range(
                cursor,
                count,
                cells,
                "Packed token command advances beyond grid",
            )?;
            let payload = read_payload(bytes, &mut offset, token_bits)?;
            let glyph = read_bits(payload, 0, 6);
            let foreground = read_bits(payload, 6, palette_bits);
            let background = read_bits(payload, 6 + palette_bits, palette_bits);
            require_zero_padding(payload, token_bits)?;
            for index in 0..count {
                write_token(
                    &mut state,
                    cursor + index,
                    glyph,
                    foreground,
                    background,
                    palette_depth,
                )?;
            }
            cursor += count;
            continue;
        }

        let repeating = matches!(
            opcode,
            REPEAT_GLYPH | REPEAT_FOREGROUND | REPEAT_BACKGROUND | REPEAT_COLOR_PAIR
        );
        let count = if repeating {
            read_count(bytes, &mut offset)?
        } else {
            1
        };
        require_range(
            cursor,
            count,
            cells,
            "Packed component command advances beyond grid",
        )?;
        let glyph = matches!(opcode, SET_GLYPH | REPEAT_GLYPH);
        let pair = matches!(opcode, SET_COLOR_PAIR | REPEAT_COLOR_PAIR);
        if !matches!(
            opcode,
            SET_GLYPH
                | SET_FOREGROUND
                | SET_BACKGROUND
                | SET_COLOR_PAIR
                | REPEAT_GLYPH
                | REPEAT_FOREGROUND
                | REPEAT_BACKGROUND
                | REPEAT_COLOR_PAIR
        ) {
            return Err(format!("Unknown mandatory packed opcode 0x{opcode:02x}"));
        }
        let used_bits = if glyph {
            6
        } else if pair {
            palette_bits * 2
        } else {
            palette_bits
        };
        let payload = read_payload(bytes, &mut offset, used_bits)?;
        let first = read_bits(payload, 0, if glyph { 6 } else { palette_bits });
        let second = pair.then(|| read_bits(payload, palette_bits, palette_bits));
        require_zero_padding(payload, used_bits)?;
        if glyph {
            if first >= 64 {
                return Err("Glyph index exceeds canonical set".to_owned());
            }
        } else if first >= palette_depth || second.is_some_and(|value| value >= palette_depth) {
            return Err("Palette index exceeds declared depth".to_owned());
        }
        for index in 0..count {
            let target = (cursor + index) * 3;
            match opcode {
                SET_GLYPH | REPEAT_GLYPH => state[target] = first as u8,
                SET_FOREGROUND | REPEAT_FOREGROUND => state[target + 1] = first as u8,
                SET_BACKGROUND | REPEAT_BACKGROUND => state[target + 2] = first as u8,
                SET_COLOR_PAIR | REPEAT_COLOR_PAIR => {
                    state[target + 1] = first as u8;
                    state[target + 2] = second.expect("color-pair opcodes set a second value") as u8;
                }
                _ => unreachable!(),
            }
        }
        cursor += count;
    }
    if !ended {
        return Err("Packed command stream has no END".to_owned());
    }
    if offset != bytes.len() {
        return Err("Trailing bytes after packed frame END".to_owned());
    }
    Ok(state)
}

fn palette_bits(depth: usize) -> Result<usize, String> {
    if !(2..=256).contains(&depth) {
        return Err("Palette depth must be an integer from 2 through 256".to_owned());
    }
    Ok(usize::BITS as usize - (depth - 1).leading_zeros() as usize)
}

fn read_count(bytes: &[u8], offset: &mut usize) -> Result<usize, String> {
    let value = read_varuint(bytes, offset)?;
    if value == 0 {
        return Err("Zero-progress command".to_owned());
    }
    usize::try_from(value).map_err(|_| "Command count exceeds platform range".to_owned())
}

fn require_range(cursor: usize, count: usize, cells: usize, message: &str) -> Result<(), String> {
    if cursor.checked_add(count).is_none_or(|end| end > cells) {
        return Err(message.to_owned());
    }
    Ok(())
}

fn read_payload<'a>(bytes: &'a [u8], offset: &mut usize, bits: usize) -> Result<&'a [u8], String> {
    let length = bits
        .checked_add(7)
        .ok_or_else(|| "Packed payload length overflow".to_owned())?
        / 8;
    let end = offset
        .checked_add(length)
        .ok_or_else(|| "Packed payload range overflow".to_owned())?;
    let payload = bytes
        .get(*offset..end)
        .ok_or_else(|| "Truncated packed command payload".to_owned())?;
    *offset = end;
    Ok(payload)
}

fn read_bits(bytes: &[u8], start: usize, width: usize) -> usize {
    let mut value = 0usize;
    for bit in 0..width {
        let absolute = start + bit;
        if bytes[absolute >> 3] & (1 << (absolute & 7)) != 0 {
            value |= 1usize << bit;
        }
    }
    value
}

fn require_zero_padding(bytes: &[u8], used_bits: usize) -> Result<(), String> {
    for bit in used_bits..bytes.len() * 8 {
        if bytes[bit >> 3] & (1 << (bit & 7)) != 0 {
            return Err("Nonzero packed padding bits".to_owned());
        }
    }
    Ok(())
}

fn write_token(
    state: &mut [u8],
    cell: usize,
    glyph: usize,
    foreground: usize,
    background: usize,
    palette_depth: usize,
) -> Result<(), String> {
    if glyph >= 64 {
        return Err("Glyph index exceeds canonical set".to_owned());
    }
    if foreground >= palette_depth || background >= palette_depth {
        return Err("Palette index exceeds declared depth".to_owned());
    }
    let target = cell
        .checked_mul(3)
        .ok_or_else(|| "Cell offset overflow".to_owned())?;
    state[target] = glyph as u8;
    state[target + 1] = foreground as u8;
    state[target + 2] = background as u8;
    Ok(())
}

fn read_varuint(bytes: &[u8], offset: &mut usize) -> Result<u32, String> {
    let start = *offset;
    let mut value = 0u32;
    let mut shift = 0u32;
    for index in 0..5usize {
        let byte = bytes
            .get(*offset)
            .copied()
            .ok_or_else(|| "Truncated varuint".to_owned())?;
        *offset += 1;
        if index == 4 && byte & 0xf0 != 0 {
            return Err("Varuint exceeds uint32".to_owned());
        }
        value |= u32::from(byte & 0x7f) << shift;
        if byte & 0x80 == 0 {
            let canonical = match value {
                0..=0x7f => 1,
                0x80..=0x3fff => 2,
                0x4000..=0x1f_ffff => 3,
                0x20_0000..=0x0fff_ffff => 4,
                _ => 5,
            };
            if *offset - start != canonical {
                return Err("Non-canonical varuint".to_owned());
            }
            return Ok(value);
        }
        shift += 7;
    }
    Err("Varuint exceeds five bytes".to_owned())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn decodes_component_updates() {
        let prior = vec![1, 2, 3, 4, 5, 6];
        let commands = [SET_GLYPH, 7, SET_COLOR_PAIR, 0x98, END];
        let state = apply_packed_commands(&commands, Some(&prior), 2, 1, 16, false)
            .expect("component updates should decode");
        assert_eq!(state, vec![7, 2, 3, 4, 8, 9]);
    }

    #[test]
    fn rejects_oversized_grid_before_allocation() {
        assert!(apply_packed_commands(&[END], None, usize::MAX, 2, 16, true).is_err());
        assert!(apply_packed_commands(&[END], None, 513, 1, 16, true).is_err());
    }
}
